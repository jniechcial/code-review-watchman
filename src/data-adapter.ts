export interface IDataAdapterInterface {
    fetchData(fromDate: Date, endDate: Date): Promise<any>;
}
