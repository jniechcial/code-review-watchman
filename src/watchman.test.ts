import Watchman from './watchman';
import * as DataAdapter from './data-adapter';

const mockedData = [
    {
        id: '1',
        url: '/test/1',
        user: { login: 'test1' },
        reviews: [{
            user: { login: 'test2' },
            state: 'APPROVED',
        }],
    },
    {
        id: '1',
        url: '/test/1',
        user: { login: 'test2' },
        reviews: [{
            user: { login: 'test3' },
            state: 'APPROVED',
        }],
    },
];

class MockedDataAdapter implements DataAdapter.IDataAdapterInterface {
    spy: any;

    constructor(spy : any) {
        this.spy = spy;
    }

    fetchData(fromDate: Date, endDate: Date): Promise<any> {
        this.spy();
        return Promise.resolve(mockedData);
    }
}

describe('Watchman', () => {
    it('#getUsers creates users repository by fetching data from dataAdapter', async () => {
        const fetchDataSpy = jest.fn();

        const mockedDataAdapter = new MockedDataAdapter(fetchDataSpy);
        const watchman = new Watchman(new Date(), new Date(), mockedDataAdapter);
        watchman.getUsers().then((usersRepository) => {
            expect(fetchDataSpy).toHaveBeenCalled();
            expect(usersRepository.getUser('test1').username).toEqual('test1');
            expect(usersRepository.getUser('test2').username).toEqual('test2');
            expect(usersRepository.getUser('test1').pullRequestsReviewed.length).toEqual(0);
            expect(usersRepository.getUser('test1').pullRequestsSubmitted.length).toEqual(1);
        });
    });
});
