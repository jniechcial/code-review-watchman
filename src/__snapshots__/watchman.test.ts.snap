// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Watchman #getUsers creates users repository by fetching data from dataAdapter 1`] = `
User {
  "pullRequestsReviewed": Array [],
  "pullRequestsSubmitted": Array [
    PullRequest {
      "id": "1",
      "reviews": Array [
        PullRequestReview {
          "state": "APPROVED",
          "username": "test2",
        },
      ],
      "url": "/test/1",
    },
  ],
  "username": "test1",
}
`;

exports[`Watchman #getUsers creates users repository by fetching data from dataAdapter 2`] = `
User {
  "pullRequestsReviewed": Array [
    PullRequest {
      "id": "1",
      "reviews": Array [
        PullRequestReview {
          "state": "APPROVED",
          "username": "test2",
        },
      ],
      "url": "/test/1",
    },
  ],
  "pullRequestsSubmitted": Array [
    PullRequest {
      "id": "1",
      "reviews": Array [
        PullRequestReview {
          "state": "APPROVED",
          "username": "test3",
        },
      ],
      "url": "/test/1",
    },
  ],
  "username": "test2",
}
`;
